README.md
<!--
SPDX-License-Identifier: CC-PDDC
SPDX-FileCopyrightText: Public domain
-->
# ATLAST-96

ATLAST-32 1.2 + ATLAST-64 2.0 Unification


ChangeLog
                         ATLAST Development Log

                             by John Walker

2007 October 7

* Changed the development version to 1.2.

* Added a new READONLYSTRINGS package configuration definition which
  causes the names of primitive words to be copied to a monolithic
  buffer in which WORDSUSED may tag them when used without running
  afoul of GCC's fascist enforcement of read-only strings. Note that
  even if READONLYSTRINGS is defined, the copy of the primitive names
  is made only if WORDSUSED is also defined.

2007 October 9

* Created a new "dist" target which builds the distribution archive
  automatically, containing only the files it is supposed to include,
  regardless of the contents of the development directory.

* Updated the MANIFEST to reflect files included in the current
  distribution.

* Removed the obsolete RTF version of the manual.

* Replaced the PostScript version of the LaTeX manual with a PDF
  edition.

* Removed READONLYSTRINGS from the package configuration section and
  made it a global configuration parameter, set by default from the
  C compiler command-line options in the Makefile.  It is not really
  a package and fits better as a global configuration item, which is
  how it is now documented.

* Completed the transmogrification of the manual into XHTML 1.0
  Strict.

* Retrofitted substantive changes to the LaTeX manual and rebuilt
  the PDF.

2007 October 10

* Removed the README from the MANIFEST and distribution; it is
  replaced by the XHTML documentation.  A place-holder README will be
  placed in the Web tree which directs anybody who links to it to the
  Web documentation.

* Modified the command line processing code in atlmain.c to dispense
  with an erroneous "uninitialized variable" natter from GCC 4.1.2
  in -O3 -Wall mode.

* Modified P_strreal in atlast.c to use a union to place the scanned
  floating point number onto the stack instead of the cast pointer
  which sufficed for the last 17 years.  GCC 4.1.2 not only generates
  a de-referencing type-punned pointer will break strict-aliasing
  rules warning in -O3 -Wall mode, it actually generates incorrect
  code for this case.

* Modified the TokReal case of atl_eval in atlast.c to use a union to
  store a scanned real number token onto the compiled output (compile
  state) or the stack (execution state).  This avoids the GCC natter
  mentioned above which, in this case, did not generate incorrect code
  with the -O3 option.

* If READONLYSTRINGS was defined, atl_primdef() in atlast.c failed
  to correctly add the names of user-defined primitives to the
  dictionary because it erroneously referenced the built-in primitive
  table (primt) rather than the primitive table passed as the
  argument (pt)--fixed.

* The sample main program in the "Integrating Atlast" section of the
  manual would generate a compiler warning on GCC with -Wall because
  the main program was not declared as returning an int--fixed.

* Added a new primdeftest.c program, which tests definition of the
  TIME and HHMMSS primitives as documented in the manual, to the
  MANIFEST and source distribution.

2014 July 3

* Changed development version number to 2.0.

* Completed initial port to 64-bit architecture.  The 64-bit
  implementation uses a different (and simpler) memory architecture
  which takes advantage of the fact that integers, floating point
  numbers, and pointers are all the same 64 bit length.  Given the
  major changes all over the program, I decided not to use
  conditional compilation but rather diverge the source code
  stream for the 64-bit version.  Because floating point values
  are now the same size of integers, programs which use floating
  point are not source code compatible with programs for the original
  32-bit version.  For example, if you wanted to duplicate the top
  floating point value on the stack, you previously had to use
  2DUP since it occupied two stack items, but now you must use
  DUP, as it now uses just one.

* I made the corresponding changes to the regression test, which
  were major since not only did many floating point tests need to
  be updated, but also the tests for the low-level stack and heap
  allocation required modification to accommodate the 8 byte size
  of stack items.  The regression test now runs cleanly.

* I removed the ALIGNMENT option from the default build in the
  Makefile.  With all stack items now being 8 bytes, there should
  be no need for this on at least the x86_64 architecture most
  people will be using.  If you are building for an architecture
  which does require alignment, you can define the tag: it still
  works.

2014 July 4

* Rebuilt the regression test expected output file, regout.txt, to
  accommodate changes due to the new memory architecture.

2014 July 5

* I decided that supporting the ANS FORTH 94 floating point primitive
  names was wasteful and not in the spirit of Atlast, especially since
  we can now take advantage of the fact that integers, pointers, and
  real numbers are all the same length.  I removed the synonym
  definitions and updated the regression test accordingly.

* Completed a port of the floating point benchmark to Atlast.  This
  is based upon the original Gforth version.

* Updated the XHTML documentation.  The changes consequent to the new
  memory architecture were major.  I also added documentation for the
  EVALUATE package and word, which has been in there for ages but did
  not appear in the documentation.

* A few typographical errors in the sample programs were corrected.

2014 July 6

* Completely reorganized the development and distribution directory
  structure.  There are now sub-directories for benchmark (the Atlast
  vs. C square root benchmark), examples (examples from the manual
  plus the floating point benchmark), and regression (the regression
  test and expected output).

* Updated the Makefile to create the new distribution archive format.


atlast.h
/*
 * vim: filetype=c:tabstop=2:ai:expandtab
 * SPDX-License-Identifier: CC-PDDC
 * SPDX-FileCopyrightText: Public domain
 *
 * A T L A S T
 * Autodesk Threaded Language Application System Toolkit
 *
 * Program Linkage Definitions
 *
 * Designed and implemented in January of 1990 by John Walker
 */

/*
 * This module contains the definitions needed by programs that invoke the
 * ATLAST system.  It also contains the definitions used internally within
 * ATLAST (which might create conflicts if included in calling programs).
 */

#define ATLAST_VERSION  "20230124-next"

#ifdef ATLAST_64BIT
# include <stdint.h>
  typedef int64_t atl_int;  /* 64-bit stack integer type */
#else
  typedef long atl_int;     /* 32-bit stack integer type */
#endif /* ifdef ATLAST_64BIT */

typedef double atl_real;    /* Real number type */

/*
 * External symbols accessible by the calling program.
 */

extern atl_int  atl_stklen;    /* Init / current stack length        */
extern atl_int  atl_rstklen;   /* Init / current return stack length */
extern atl_int  atl_heaplen;   /* Init / current heap length         */
extern atl_int  atl_ltempstr;  /* Temporary string buffer length     */
extern atl_int  atl_ntempstr;  /* Number of temporary string buffers */
extern atl_int  atl_trace;     /* Trace mode                         */
extern atl_int  atl_walkback;  /* Error walkback enabled mode        */
extern atl_int  atl_comment;   /* Currently ignoring comment         */
extern atl_int  atl_redef;     /* Allow redefining words without     */
                               /*   issuing the "not unique" warning */
extern atl_int  atl_errline;   /* Line number of the last atl_load() */
                               /*   error, or zero (if no error)     */

/*
 * ATL_EVAL return status codes
 */

#define ATL_SNORM           0  /* Normal evaluation                  */
#define ATL_STACKOVER      -1  /* Stack overflow                     */
#define ATL_STACKUNDER     -2  /* Stack underflow                    */
#define ATL_RSTACKOVER     -3  /* Return stack overflow              */
#define ATL_RSTACKUNDER    -4  /* Return stack underflow             */
#define ATL_HEAPOVER       -5  /* Heap overflow                      */
#define ATL_BADPOINTER     -6  /* Pointer outside the heap           */
#define ATL_UNDEFINED      -7  /* Undefined word                     */
#define ATL_FORGETPROT     -8  /* Attempt to forget protected word   */
#define ATL_NOTINDEF       -9  /* Compiler word outside definition   */
#define ATL_RUNSTRING     -10  /* Non-terminated string              */
#define ATL_RUNCOMM       -11  /* Non-terminated comment in file     */
#define ATL_BREAK         -12  /* Asynchronous break signal received */
#define ATL_DIVZERO       -13  /* Attempt to divide by zero          */
#define ATL_APPLICATION   -14  /* Application primitive atl_error()  */

/*
 * Entry points
 */

extern void atl_init(), atl_mark(), atl_unwind(), atl_break();
extern int  atl_eval(), atl_load();
extern void atl_memstat();


atldef.h
/*
 * vim: filetype=c:tabstop=2:ai:expandtab
 * SPDX-License-Identifier: CC-PDDC
 * SPDX-FileCopyrightText: Public domain
 *
 * A T L A S T
 * Autodesk Threaded Language Application System Toolkit
 *
 * Definitions for components of ATLAST
 *
 * Designed and implemented in January of 1990 by John Walker
 */

/*
 * This file contains the definitions for modules within the ATLAST
 * language system itself.  Definitions used by programs that link to
 * ATLAST are in "atlast.h", which is included by this file automatically.
 */

#include "atlast.h" /* Define user linkage structures */

typedef void (*codeptr) (); /* Machine code pointer */

/*
 * Dictionary word entry
 */

typedef struct dw
{
  struct dw *wnext;  /* Next word in dictionary.             */
  char *wname;       /* Word name.  The first character is   */
                     /*   actually the word flags, including */
                     /*   the (IMMEDIATE) bit.               */
  codeptr wcode;     /* Machine code implementation.         */
} dictword;

/*
 * Word flag bits
 */

#define IMMEDIATE    1  /* Word is immediate          */
#define WORDUSED     2  /* Word used by program       */
#define WORDHIDDEN   4  /* Word is hidden from lookup */

/*
 * Data types
 */

#ifdef ATLAST_64BIT
  typedef atl_int stackitem;
#else
  typedef long stackitem;
#endif /* ifdef ATLAST_64BIT */

typedef dictword **rstackitem;

/*
 * Stack items occupied by a dictionary word definition
 */

#define Dictwordl                  \
   (( sizeof ( dictword )          \
  + ( sizeof ( stackitem ) - 1 ))  \
    / sizeof ( stackitem ))

/*
 * Token types
 */

#define TokNull     0  /* Nothing scanned                  */
#define TokWord     1  /* Word stored in token name buffer */
#define TokInt      2  /* Integer scanned                  */
#define TokReal     3  /* Real number scanned              */
#define TokString   4  /* String scanned                   */

/*
 * Primitive definition table entry
 */

struct primfcn
{
  char *pname;
  codeptr pcode;
};

/*
 * Internal state marker item
 */

typedef struct
{
  stackitem *mstack;    /* Stack position marker        */
  stackitem *mheap;     /* Heap allocation marker       */
  dictword ***mrstack;  /* Return stack position marker */
  dictword *mdict;      /* Dictionary marker            */
} atl_statemark;

#ifdef EXPORT
# define Exported
# ifndef NOMANGLE
#  define stk          atl__sp
#  define stack        atl__sk
#  define stackbot     atl__sb
#  define stacktop     atl__st
#  define rstk         atl__rp
#  define rstack       atl__rk
#  define rstackbot    atl__rb
#  define rstacktop    atl__rt
#  define heap         atl__hb
#  define hptr         atl__hp
#  define heapbot      atl__hs
#  define heaptop      atl__ht
#  define dict         atl__dh
#  define dictprot     atl__dp
#  define strbuf       atl__ts
#  define cstrbuf      atl__tn
#  define ip           atl__ip
#  define curword      atl__cw
#  define createword   atl__wd
# endif /* ifndef NOMANGLE */

# ifndef NOMANGLE
#  define stackmax    atl__sx
#  define rstackmax   atl__rx
#  define heapmax     atl__hx
# endif /* ifndef NOMANGLE */
    extern stackitem *stackmax, *heapmax;
    extern dictword ***rstackmax;

# ifdef ALIGNMENT
#  ifndef NOMANGLE
#   define rbuf0   atl__r0
#   define rbuf1   atl__r1
#   define rbuf2   atl__r2
#  endif /* ifndef NOMANGLE */
    extern atl_real rbuf0, rbuf1, rbuf2; /* Real temporaries for alignment */
# endif /* ifdef ALIGNMENT */

# define FmodeR     1  /* Read mode        */
# define FmodeW     2  /* Write mode       */
# define FmodeB     4  /* Binary file mode */
# define FmodeCre   8  /* Create new file  */

  extern stackitem *   stack, *stk, *stackbot, *stacktop,
                       *heap, *hptr, *heapbot, *heaptop;
  extern dictword ***  rstack, ***rstk, ***rstackbot, ***rstacktop;
  extern dictword *    dict, *dictprot, *curword, *createword;
  extern dictword **   ip;
  extern char **       strbuf;
  extern int           cstrbuf;

# ifndef NOMANGLE
#  define P_create   atl__Pcr
#  define P_dodoes   atl__Pds
# endif /* ifndef NOMANGLE */
  extern void P_create(), P_dodoes();
#else
# define Exported  static
#endif /* ifdef EXPORT */

#ifndef NOMEMCHECK
# ifdef EXPORT
#  ifndef NOMANGLE
#   define stakover     atl__Eso
#   define rstakover    atl__Erso
#   define heapover     atl__Eho
#   define badpointer   atl__Ebp
#   define stakunder    atl__Esu
#   define rstakunder   atl__Ersu
#  endif /* ifdef NOMANGLE */
    extern
# endif /* ifdef EXPORT */
  void
  stakover(),
  rstakover(), heapover(), badpointer(), stakunder(), rstakunder();
#endif /* ifndef NOMEMCHECK */

/*
 * Functions called by exported extensions.
 */

extern void atl_primdef(), atl_error();
extern dictword *atl_lookup(), *atl_vardef();
extern stackitem *atl_body();
extern int atl_exec();
#ifdef EXPORT
  extern char *atl_fgetsp();
#endif /* ifdef EXPORT */

/*
 * If explicit alignment is not requested, enable it in
 * any case for known CPU types that require alignment.
 */

#ifndef ALIGNMENT
# ifdef sparc
#  define ALIGNMENT
# endif /* ifdef sparc */
#endif /* ifndef ALIGNMENT */

/*
 * STATIC is defined as "static" for platforms which require this
 * specification on declarations for forward-referenced functions.
 */

#define STATIC  static

/*
 * Dynamic storage manipulation primitives
 */

/*
 * Stack access definitions
 */

#define S0        stk[-1]        /* Top of stack                   */
#define S1        stk[-2]        /* Next on stack                  */
#define S2        stk[-3]        /* Third on stack                 */
#define S3        stk[-4]        /* Fourth on stack                */
#define S4        stk[-5]        /* Fifth on stack                 */
#define S5        stk[-6]        /* Sixth on stack                 */
#define Pop       stk--          /* Pop the top item off the stack */
#define Pop2      stk  -= 2      /* Pop two items off the stack    */
#define Npop(n)   stk  -= ( n )  /* Pop N items off the stack      */
#define Push      *stk++         /* Push item onto stack           */

# define Mss(n)                     \
  if (( stk + ( n )) > stackmax)    \
  stackmax = stk + ( n );
# define Msr(n)                     \
  if (( rstk + ( n )) > rstackmax)  \
  rstackmax = rstk + ( n );
# define Msh(n)                     \
  if (( hptr + ( n )) > heapmax)    \
  heapmax = hptr + ( n );

#ifdef NOMEMCHECK
# define Sl(x)
# define So(n)
#else
# define Memerrs
# define Sl(x)                           \
  if (( stk - stack ) < ( x ))           \
    {                                    \
      stakunder();                       \
      return Memerrs;                    \
    }
# define So(n)                           \
  Mss(n) if (( stk + ( n )) > stacktop)  \
    {                                    \
      stakover();                        \
      return Memerrs;                    \
    }
#endif /* ifdef NOMEMCHECK */

/*
 * Return stack access definitions
 */

#define R0      rstk[-1]  /* Top of return stack   */
#define R1      rstk[-2]  /* Next on return stack  */
#define R2      rstk[-3]  /* Third on return stack */
#define Rpop    rstk--    /* Pop return stack      */
#define Rpush   *rstk++   /* Push return stack     */
#ifdef NOMEMCHECK
# define Rsl(x)
# define Rso(n)
#else
# define Rsl(x)                            \
  if (( rstk - rstack ) < ( x ))           \
    {                                      \
      rstakunder();                        \
      return Memerrs;                      \
    }
# define Rso(n)                            \
  Msr(n) if (( rstk + ( n )) > rstacktop)  \
    {                                      \
      rstakover();                         \
      return Memerrs;                      \
    }
#endif /* ifdef NOMEMCHECK */

/*
 * Heap access definitions
 */

#ifdef NOMEMCHECK
# define Ho(n)
# define Hpc(n)
#else
# define Ho(n)                             \
  Msh(n) if (( hptr + ( n )) > heaptop)    \
    {                                      \
      heapover();                          \
      return Memerrs;                      \
    }
# define Hpc(n)                            \
  if ((((stackitem *)( n )) <  heapbot )   \
   || (((stackitem *)( n )) >= heaptop ))  \
    {                                      \
      badpointer();                        \
      return Memerrs;                      \
    }
#endif /* ifdef NOMEMCHECK */

#define Hstore   *hptr++      /* Store item on heap                 */
#define state    ( *heap )    /* Execution state is first heap word */
#define prim     static void  /* Attributes for primitive functions */

/*
 * Real number definitions
 * (used only if REAL is configured).
 */

#define Realsize \
  ( sizeof ( atl_real ) / sizeof ( stackitem )) /* Stack cells / real       */
#define Realpop    stk  -= Realsize             /* Pop real from stack      */
#define Realpop2   stk  -= ( 2 * Realsize )     /* Pop two reals from stack */

#ifdef ATLAST_64BIT
# ifdef ALIGNMENT
#  define REAL0 *((atl_real *) memcpy((char *) &rbuf0, (char *) &S0, sizeof(atl_real)))
#  define REAL1 *((atl_real *) memcpy((char *) &rbuf1, (char *) &S1, sizeof(atl_real)))
#  define REAL2 *((atl_real *) memcpy((char *) &rbuf2, (char *) &S2, sizeof(atl_real)))
#  define SREAL0(x) rbuf2=(x); (void)memcpy((char *) &S0, (char *) &rbuf2, sizeof(atl_real))
#  define SREAL1(x) rbuf2=(x); (void)memcpy((char *) &S1, (char *) &rbuf2, sizeof(atl_real))
# else
#  define REAL0   *((atl_real *) &S0)   /* First real on stack  */
#  define REAL1   *((atl_real *) &S1)   /* Second real on stack */
#  define REAL2   *((atl_real *) &S2)   /* Third real on stack  */
#  define SREAL0(x) *((atl_real *) &S0) = (x)
#  define SREAL1(x) *((atl_real *) &S1) = (x)
# endif
#else
# ifdef ALIGNMENT
#  define REAL0 *((atl_real *) memcpy((char *) &rbuf0, (char *) &S1, sizeof(atl_real)))
#  define REAL1 *((atl_real *) memcpy((char *) &rbuf1, (char *) &S3, sizeof(atl_real)))
#  define REAL2 *((atl_real *) memcpy((char *) &rbuf2, (char *) &S5, sizeof(atl_real)))
#  define SREAL0(x) rbuf2=(x); (void)memcpy((char *) &S1, (char *) &rbuf2, sizeof(atl_real))
#  define SREAL1(x) rbuf2=(x); (void)memcpy((char *) &S3, (char *) &rbuf2, sizeof(atl_real))
# else
#  define REAL0   *((atl_real *) &S1)   /* First real on stack  */
#  define REAL1   *((atl_real *) &S3)   /* Second real on stack */
#  define REAL2   *((atl_real *) &S5)   /* Third real on stack  */
#  define SREAL0(x) *((atl_real *) &S1) = (x)
#  define SREAL1(x) *((atl_real *) &S3) = (x)
# endif
#endif /* ifdef ATLAST_64BIT */

/*
 * File I/O definitions
 * (used only if FILEIO is configured)
 */

#define FileSent 0x831FDF9DL

#define Isfile(x)                             \
  Hpc(x);                                     \
  if (*((stackitem *)( x )) != FileSent)      \
    {                                         \
      V printf("\nNot a file\n");             \
      return;                                 \
    }
#define FileD(x)                              \
      ((FILE *)*(((stackitem *)( x )) + 1 ))
#define Isopen(x)                             \
  if (FileD(x) == NULL)                       \
    {                                         \
      V printf("\nFile not open\n");          \
      return;                                 \
    }


primdeftest.c
/*
 * vim: filetype=c:tabstop=2:ai:expandtab
 * SPDX-License-Identifier: CC-PDDC
 * SPDX-FileCopyrightText: Public domain
 *
 * A T L A S T
 * PRIMDEFTEST
 *
 * Test primitive definition in ATLAST
 *
 * Designed and implemented in January of 1990 by John Walker
 */

#include <stdio.h>
#include <stdlib.h>
#include <ctype.h>
#include <string.h>
#include <time.h>

#include "atldef.h"

prim
ptime()
{
  So(1);
  Push = time(NULL);
}

prim
phhmmss()
{
  struct tm *lt;

  Sl(1);
  So(2);
  lt    = localtime(&S0);
  S0    = lt->tm_hour;
  Push  = lt->tm_min;
  Push  = lt->tm_sec;
}

prim
pleibniz()
{
  long    nterms;
  double   sum = 0.0;
  double numer = 1.0;
  double denom = 1.0;

  Sl(1);
  nterms = S0;
  Pop;

  So(Realsize);
  Push = 0;
  while (nterms-- > 0)
    {
      sum    += numer / denom;
      numer   = -numer;
      denom  += 2.0;
    }
  SREAL0(sum * 4.0);
}

static struct primfcn timep[] = { { "0TIME",    ptime      },
                                  { "0HHMMSS",  phhmmss    },
                                  { "0LEIBNIZ", pleibniz   },
                                  { NULL,       (codeptr)0 } };

int
main()
{
  char t[132];

  atl_init();
  atl_primdef(timep);
  while (fprintf(stdout, "-> "), fgets(t, 132, stdin) != NULL) /* //-V639 */
    {
      atl_eval(t);
    }
  return 0;
}


atlmain.c
/*
 * vim: filetype=c:tabstop=2:ai:expandtab
 * SPDX-License-Identifier: CC-PDDC
 * SPDX-FileCopyrightText: Public domain
 *
 * A T L A S T
 * ATLMAIN
 *
 * Main driver program for interactive ATLAST
 *
 * Designed and implemented in January of 1990 by John Walker
 */

#include <stdio.h>
#include <stdlib.h>
#include <ctype.h>
#include <string.h>
#include <signal.h>

#include "atldef.h"

#define FALSE   0
#define TRUE    1
#define V       (void)

/*
 * Globals imported
 */

/*
 * CTRLC
 * Catch a user console break signal.
 *
 * If your C library does not provide this POSIX / UNIX-compatible facility
 * (registered with the call on signal() in main()), just turn this code
 * off or, better still, replace it with the equivalent on your system.
 */

  static void ctrlc(sig) int sig;
  {
    if (sig == SIGINT)
      {
        atl_break();
      }
  }

/*
 * MAIN
 * Main program
 */

int
main(argc, argv)
int argc;
char *argv[];
{
  int    i;
  int    fname   = FALSE;
  int    defmode = FALSE;
  FILE * ifp;
  char * include[20];
  int    in = 0;

  include[1] = NULL;

#define PR(x)  ((void)fprintf(stderr, (x)))

#ifdef ATLAST_64BIT
    PR("\rATLAST 96." ATLAST_VERSION " 64-bit (Built " __DATE__")\r\n");
#else
    PR("\rATLAST 96." ATLAST_VERSION " 32-bit (Built " __DATE__")\r\n");
#endif /* ifdef ATLAST_64BIT */
  ifp = stdin;
  for (i = 1; i < argc; i++)
    {
      char *cp, opt;

      cp = argv[i];
      if (*cp == '-')
        {
          opt = *( ++cp );
          if (islower(opt))
            {
              opt = toupper(opt);
            }

          switch (opt)
            {
            case 'D':
              defmode = TRUE;
              break;

            case 'H':
              atl_heaplen = atol(cp + 1);
              break;

            case 'I':
              include[in++] = cp + 1;
              break;

            case 'R':
              atl_rstklen = atol(cp + 1);
              break;

            case 'S':
              atl_stklen = atol(cp + 1);
              break;

            case 'T':
              atl_trace = TRUE;
              break;

            case '?':
            case 'U':
              PR("Usage:  atlast [options] [inputfile]\n");
              PR("        Options:\n");
              PR("           -d        Treat file as definitions\n");
              PR("           -h<n>     Heap length <n>\n");
              PR("           -i<file>  Include named definition <file>\n");
              PR("           -r<n>     Return stack length <n>\n");
              PR("           -s<n>     Stack length <n>\n");
              PR("           -t        Set trace mode\n");
              PR("           -u, -?    Print this message\n");
              return 0;
            }
        }
      else
        {
          char fn[132];

          if (fname)
            {
              PR("Duplicate file name.\n");
              return 1;
            }

          fname = TRUE;
          V strcpy(fn, cp);
          if (strchr(fn, '.') == NULL)
            {
              V strcat(fn, ".atl");
            }

          ifp = fopen(fn, "r");
          if (ifp == NULL)
            {
              V fprintf(stderr, "Unable to open file %s\n", fn);
              return 1;
            }
        }
    }

  /*
   * If any include files were named, load each
   * in turn before we execute the program.
   */

  for (i = 0; i < in; i++)
    {
      int    stat;
      char   fn[132];
      FILE * fp;

      V strcpy(fn, include[i]);
      if (strchr(fn, '.') == NULL)
        {
          V strcat(fn, ".atl");
        }

      fp = fopen(fn, "r");
      if (fp == NULL)
        {
          V fprintf(stderr, "Unable to open include file %s\n", include[i]);
          return 1;
        }

      stat = atl_load(fp);
      V fclose(fp);
      if (stat != ATL_SNORM)
        {
          V printf("\nError %d in include file %s\n", stat, include[i]);
        }
    }

  /*
   * Now that all the preliminaries are out of the way,
   * fall into the main ATLAST execution loop.
   */

  V signal(SIGINT, ctrlc);

  while (TRUE)
    {
      char t[132];

      if (!fname)
        {
          V printf(atl_comment ? "(  " :  /* Show pending comment */
                                          /* Show compiling state */
                   ((( heap != NULL ) && state ) ? ":> " : "-> " ));
        }

      if (fgets(t, 132, ifp) == NULL)
        {
          if (fname && defmode)
            {
              fname  = defmode = FALSE;
              ifp    = stdin;
              continue;
            }

          break;
        }

      V atl_eval(t);
    }
  if (!fname)
    {
      V printf("\n");
    }

  return 0;
}
